### PR TITLE
ir: Make projections local to a `Body`

### LINKED ISSUE #996: ir+lower: consider localising `Projections` to each body
## Description

Currently, projections are interned in the global IR projection store. However, there is no need for the items to be all in a global store since the projections have no relation to one another beyond the scope of a `Body`. A potential change could be introduced to make all local projections be stored on the `Body`.

## Why do this?

In the future, we want to parallelise IR building and optimising as much as possible. The thing that hurts the most is the contention on locking (even if we are just reading it, there is still overhead). In order to completely avoid locking when reading projections, we could just store them local to a body, and read them with the owning `Body` in scope.

## Required changes
- Change the IR to store local projections on the `Body`
- Change lowering to use and create projections on the built body. 
- Change the IR visitors to access projections from the owning `Body`

### PR DESCRIPTION
**Pull request by feds01**
This PR makes all of the projections that are applied to a body local to the `Body` rather than being externally stored in a static store. The implementation isn't as clean as I'd hoped it to be (especially in the visitor parts).

## Changes

- ir: localise `Projections` to each `Body`
- ir: cleanup `IrVisitorMut` with `VisitorCtx`
- ir: cleanup `ModiyfyingIrVisitor` with `IrVisitorCtxMut`

## Todo

- [x] Merge #1009 before reviewing this PR

closes #996

### CODE REVIEW

File: compiler/hash-storage/src/store/sequence/local.rs [hunk 1]
```diff
@@ -0,0 +1,71 @@
+use std::marker::PhantomData;
+
+use super::SequenceStoreKey;
+
+/// An implementation of [SequenceStore] which doesn't require being shared
+/// across thread boundaries.
+pub struct LocalSequenceStore<Key, V> {
```
Comment (kontheocharis) [inline compiler/hash-storage/src/store/sequence/local.rs:7]: How is this different from `IndexVec`?

Comment (feds01) [reply compiler/hash-storage/src/store/sequence/local.rs:7]: It can use the `SequenceStoreKey` API out of the box

Comment (kontheocharis) [reply compiler/hash-storage/src/store/sequence/local.rs:7]: Okay, I see. Perhaps we should slowly start using this over `IndexVec` then

File: compiler/hash-storage/src/store/sequence/local.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    /// Get the value sequence for the given key as an owned vector.
+    pub fn get_vec(&self, key: Key) -> Vec<V> {
+        let (index, len) = key.to_index_and_len();
+        self.data[index..index + len].to_vec()
+    }
+
+    /// Borrow a collection of items.
+    pub fn borrow(&self, key: Key) -> &[V] {
+        let (index, len) = key.to_index_and_len();
+        &self.data[index..index + len]
```
Comment (kontheocharis) [inline compiler/hash-storage/src/store/sequence/local.rs:57]: Can do `unchecked` here too

File: compiler/hash-storage/src/store/sequence/local.rs [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    /// Borrow a mutable collection of items.
+    pub fn borrow_mut(&mut self, key: Key) -> &mut [V] {
+        let (index, len) = key.to_index_and_len();
+        &mut self.data[index..index + len]
```
Comment (kontheocharis) [inline compiler/hash-storage/src/store/sequence/local.rs:63]: And here

File: compiler/hash-storage/src/store/sequence/local.rs (hunk 1 above)
Comment (feds01) [reply compiler/hash-storage/src/store/sequence/local.rs:7]: Definitely in areas that lend them selves to using sequence stores

**Comment by kontheocharis** [comment]
LGTM after above

**Review by kontheocharis — APPROVED** [review]

**Pull request merged**